### PR TITLE
02181 d tokendelete update fee calc

### LINKED ISSUE #2181: Standardize fee calculations for TokenDelete, TokenUpdate
### Problem

Following example of `TokenCreate` example, refactor resource usage calculations of `TokenDelete`, `TokenUpdate`, `TokenGrantKyc`, `TokenRevokeKyc` to be based on AccessorBasedUsages and UsageAccumulator.

### Solution

- [ ] Change the incremental resource usage as calculated  into a calculation supported AccessorBasedUsages and UsageAccumulator;
- [ ] Add a `FeeSchedulesTest` method to generate the prices for _feeSchedules.json_, **given** a base operation defined by `BaseOperationUsage` (and the target price in _hapi-fees/src/main/resources/canonical-prices.json_);

Reference to issue #2071. 

### Alternatives

_No response_

### LINKED ISSUE #2182: Standardize fee calculations for TokenAssociate, TokenDissociate
### Problem

Following example of `TokenCreate` example, refactor resource usage calculations of `TokenAssociate`, `TokenDissociate` to be based on AccessorBasedUsages and UsageAccumulator.

### Solution

- [ ] Change the incremental resource usage as calculated  into a calculation supported AccessorBasedUsages and UsageAccumulator;
- [ ] Add a `FeeSchedulesTest` method to generate the prices for _feeSchedules.json_, **given** a base operation defined by `BaseOperationUsage` (and the target price in _hapi-fees/src/main/resources/canonical-prices.json_);

Reference to issue #2071. 

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by ljianghedera**
**Description**:
**Related issue(s)**:

- Closes #2181 
- Closed #2182

**Checklist**

- [x] Refactors fee calculation for TokenUpdate, TokenDelete, TokenAssociate, TokenDissociate, TokenGrantKycToAccount, and TokenRevokeKycFromAccount transactions into calculation supported by AccessorBasedUsages and UsageAccumulator;
- [x] Updates BaseOperationUsage to support , with the base resource usage matching that in typedUsageFormulas.json;
- [x] Adds `TokenFeeSchedulesTest` methods to validate TokenUpdate, TokenDelete, TokenAssociate, TokenDissociate, TokenGrantKycToAccount, and TokenRevokeKycFromAccount the prices with base usage are exactly $0.001 respectively as expected;
- [x] Updates the ScheduleGenerator.SUPPORTED_FUNCTIONS list with TokenUpdate, TokenDelete, TokenAssociate, TokenDissociate, TokenGrantKycToAccount, and TokenRevokeKycFromAccount;
- [x] Replaces TokenUpdate, TokenDelete, TokenAssociate, TokenDissociate, TokenGrantKycToAccount, and TokenRevokeKycFromAccount prices in feeSchedules.json;
- [x] Updates the EET clients feeFor() function of HapiTokenUpdate, HapiTokenDelete,TokenAssociate, TokenDissociate, HapiTokenKycGrant, and HapiTokenRevokeKyc to use UsageAccumulator based resource usage classes;
- [x] Adds EET specs for TokenUpdate, TokenDelete, TokenAssociate, TokenDissociate, TokenGrantKycToAccount, and TokenRevokeKycFromAccount base operations and validate the base prices are charged as expected.
- [x] Added a new `AllTokenBaseOpFeesSuite` for token base operation fee specs;
- [x] Cleaned up legacy fee calculation code;

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hapi-fees/src/main/java/com/hedera/services/usage/token/meta/ExtantTokenContext.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+		}
+		public Builder newBuilder() {
+			return new Builder();
+		}
+
+		public ExtantTokenContext build() {
+			return new ExtantTokenContext(this);
+		}
+
+		public Builder setExistingSymLen(final int existingSymLen) {
```
Comment (MarcKriguerAtHedera) [inline hapi-fees/src/main/java/com/hedera/services/usage/token/meta/ExtantTokenContext.java:92]: for each of these 4 (or 5) methods, would it make sense to validate that these numbers are non-negative?

**Review by MarcKriguerAtHedera — APPROVED** [review]
One minor suggestion (that shows up a few other times, but it's the same comment) - otherwise Looks Good To Me.

File: hapi-fees/src/main/java/com/hedera/services/pricing/BaseOperationUsage.java [hunk 2]
```diff
@@ -111,6 +116,7 @@
 			.setEd25519(ByteString.copyFromUtf8("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"))
 			.build();
 	private static final AccountID AN_ACCOUNT = AccountID.newBuilder().setAccountNum(1_234L).build();
+	private static final TokenID AN_TOKEN = TokenID.newBuilder().setTokenNum(1_235L).build();
```
Comment (anighanta) [inline hapi-fees/src/main/java/com/hedera/services/pricing/BaseOperationUsage.java:119]: nit - typo : A_TOKEN

File: hapi-fees/src/main/java/com/hedera/services/usage/token/TokenOpsUsage.java [hunk 3]
```diff
@@ -51,6 +61,10 @@
 	private static final int ROYALTY_HBAR_FALLBACK_REPR_SIZE = ROYALTY_NO_FALLBACK_REPR_SIZE + FIXED_HBAR_REPR_SIZE;
 	private static final int ROYALTY_HTS_FALLBACK_REPR_SIZE = ROYALTY_NO_FALLBACK_REPR_SIZE + FIXED_HTS_REPR_SIZE;
 	private static final long LONG_BASIC_ENTITY_ID_SIZE = BASIC_ENTITY_ID_SIZE;
+	private static final int AMOUNT_REPR_BYTES = 8;
```
Comment (anighanta) [inline hapi-fees/src/main/java/com/hedera/services/usage/token/TokenOpsUsage.java:64]: is this long size ? I see that this is used just for `expiry`. If so we can remove this and just use `LONG_SIZE`

File: hapi-fees/src/main/java/com/hedera/services/usage/token/TokenOpsUsage.java [hunk 4]
```diff
@@ -180,7 +202,89 @@ public void tokenWipeUsage(final SigUsage sigUsage,
...
+				(extantTokenContext.getHashasAutoRenewAccount() || tokenUpdateMeta.hasAutoRenewAccount())) {
+			rbSize += BASIC_ENTITY_ID_SIZE;
+		}
 
+		rbSize += tokenUpdateMeta.getNewSymLen() > 0 ? tokenUpdateMeta.getNewSymLen() : extantTokenContext.getExistingSymLen();
+		rbSize += tokenUpdateMeta.getNewNameLen() > 0 ? tokenUpdateMeta.getNewNameLen() : extantTokenContext.getExistingNameLen();
+		rbSize += tokenUpdateMeta.getNewMemoLen() > 0 ? tokenUpdateMeta.getNewMemoLen() : extantTokenContext.getExistingMemoLen();
+
+		long newLifeTime = Math.max(tokenUpdateMeta.getNewExpiry() - tokenUpdateMeta.getNewEffectiveTxnStartTime(),
+				extantTokenContext.getExistingExpiry() - tokenUpdateMeta.getNewEffectiveTxnStartTime());
```
Comment (anighanta) [inline hapi-fees/src/main/java/com/hedera/services/usage/token/TokenOpsUsage.java:242]: could simplify as
```
long newLifeTime = Math.max(tokenUpdateMeta.getNewExpiry(), extantTokenContext.getExistingExpiry()) - tokenUpdateMeta.getNewEffectiveTxnStartTime();
```

File: hapi-fees/src/main/java/com/hedera/services/usage/token/TokenOpsUsage.java [hunk 5]
```diff
@@ -180,7 +202,89 @@ public void tokenWipeUsage(final SigUsage sigUsage,
...
+		long existingRbSize = extantTokenContext.getExistingRbSize();
+
+		long rbsDelta = Math.max(0, newLifeTime * (rbSize - existingRbSize));
+		if(rbsDelta > 0) {
+			accumulator.addRbs(rbsDelta);
+		}
+
+		long txnBytes = rbSize + BASIC_ENTITY_ID_SIZE + noRbImpactBytes(tokenUpdateMeta);
+		accumulator.addBpt(txnBytes);
+		if(tokenUpdateMeta.hasTreasure()) {
```
Comment (anighanta) [inline hapi-fees/src/main/java/com/hedera/services/usage/token/TokenOpsUsage.java:254]: typo : `hasTreasury()`

File: hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenUntypedMetaBase.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
+package com.hedera.services.usage.token.meta;
```
Comment (anighanta) [inline hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenUntypedMetaBase.java:1]: Missing license.

File: hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenRevokeKycMeta.java [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package com.hedera.services.usage.token.meta;
```
Comment (anighanta) [inline hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenRevokeKycMeta.java:1]: Missing license.

File: hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenGrantKycMeta.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package com.hedera.services.usage.token.meta;
```
Comment (anighanta) [inline hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenGrantKycMeta.java:1]: Missing license.

File: hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenDissociateMeta.java [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package com.hedera.services.usage.token.meta;
```
Comment (anighanta) [inline hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenDissociateMeta.java:1]: Missing license.

File: hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenAssociateMeta.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
+package com.hedera.services.usage.token.meta;
```
Comment (anighanta) [inline hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenAssociateMeta.java:1]: Missing license.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/token/HapiTokenAssociate.java [hunk 11]
```diff
@@ -78,17 +84,24 @@ protected HapiTokenAssociate self() {
...
-			final long expiry = lookupExpiry(spec);
-			FeeCalculator.ActivityMetrics metricsCalc = (_txn, svo) -> {
-				var estimate = TokenAssociateUsage.newEstimate(_txn, suFrom(svo))
-						.givenCurrentExpiry(expiry);
-				return estimate.get();
-			};
+			expiry = lookupExpiry(spec);
 			return spec.fees().forActivityBasedOp(
-					HederaFunctionality.TokenAssociateToAccount, metricsCalc, txn, numPayerKeys);
+					HederaFunctionality.TokenDissociateFromAccount, this::usageEstimate, txn, numPayerKeys);
```
Comment (anighanta) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/token/HapiTokenAssociate.java:89]: shouldn't this be `HederaFunctionality.TokenAssociateToAccount`

File: test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/token/HapiTokenKycRevoke.java [hunk 12]
```diff
@@ -70,9 +74,18 @@ protected long feeFor(HapiApiSpec spec, Transaction txn, int numPayerKeys) throw
...
-		return TokenRevokeKycUsage.newEstimate(txn, suFrom(svo)).get();
+		UsageAccumulator accumulator = new UsageAccumulator();
+		final var tokenRevokeKycMeta = TOKEN_OPS_USAGE_UTILS.tokenRevokeKycUsageFrom();
+		final var baseTransactionMeta = new BaseTransactionMeta(txn.getMemoBytes().size(), 0);
+		TokenOpsUsage tokenOpsUsage = new TokenOpsUsage();
+		tokenOpsUsage.tokenRevokeKycUsage(suFrom(svo), baseTransactionMeta, tokenRevokeKycMeta, accumulator );
+		return AdapterUtils.feeDataFrom(accumulator);
 	}
 
+//	private FeeData usageEstimate(TransactionBody txn, SigValueObj svo) {
```
Comment (anighanta) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/token/HapiTokenKycRevoke.java:85]: remove the commented out dead code.

**Review by anighanta — CHANGES_REQUESTED** [review]
LGTM overall. Please see inline comments.

File: hapi-fees/src/main/java/com/hedera/services/pricing/BaseOperationUsage.java (hunk 2 above)
Comment (ljianghedera) [reply hapi-fees/src/main/java/com/hedera/services/pricing/BaseOperationUsage.java:119]: Fixed.

File: hapi-fees/src/main/java/com/hedera/services/usage/token/TokenOpsUsage.java (hunk 3 above)
Comment (ljianghedera) [reply hapi-fees/src/main/java/com/hedera/services/usage/token/TokenOpsUsage.java:64]: Yeah, replaced.

File: hapi-fees/src/main/java/com/hedera/services/usage/token/TokenOpsUsage.java (hunk 4 above)
Comment (ljianghedera) [reply hapi-fees/src/main/java/com/hedera/services/usage/token/TokenOpsUsage.java:242]: Changed as suggested.

File: hapi-fees/src/main/java/com/hedera/services/usage/token/TokenOpsUsage.java (hunk 5 above)
Comment (ljianghedera) [reply hapi-fees/src/main/java/com/hedera/services/usage/token/TokenOpsUsage.java:254]: Fixed.

File: hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenUntypedMetaBase.java (hunk 6 above)
Comment (ljianghedera) [reply hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenUntypedMetaBase.java:1]: Added.

File: hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenRevokeKycMeta.java (hunk 7 above)
Comment (ljianghedera) [reply hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenRevokeKycMeta.java:1]: Added.

File: hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenGrantKycMeta.java (hunk 8 above)
Comment (ljianghedera) [reply hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenGrantKycMeta.java:1]: Added.

File: hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenAssociateMeta.java (hunk 10 above)
Comment (ljianghedera) [reply hapi-fees/src/main/java/com/hedera/services/usage/token/meta/TokenAssociateMeta.java:1]: Added.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/token/HapiTokenAssociate.java (hunk 11 above)
Comment (ljianghedera) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/token/HapiTokenAssociate.java:89]: Yes, fixed.

**Review by anighanta — DISMISSED** [review]

File: test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/token/HapiTokenKycRevoke.java (hunk 12 above)
Comment (ljianghedera) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/token/HapiTokenKycRevoke.java]: Done

File: hapi-fees/src/main/java/com/hedera/services/usage/token/meta/ExtantTokenContext.java (hunk 1 above)
Comment (ljianghedera) [reply hapi-fees/src/main/java/com/hedera/services/usage/token/meta/ExtantTokenContext.java:92]: Yeah, it makes sense to validate even though from the context these numbers won't be add. 
Added the validate logic.

**Comment by tinker-michaelj** [comment]
Will be superseded by modularization work.

**Pull request closed without merging**